packages/sources/curve/src/endpoint/crypto.ts

```diff
@@ -0,0 +1,115 @@
+import { Requester, Validator } from '@chainlink/ea-bootstrap'
+import { ExecuteWithConfig, InputParameters } from '@chainlink/types'
+import { NAME as AdapterName, Config } from '../config'
+import { ethers, BigNumber } from 'ethers'
+import addressProviderABI from '../abis/address_provider.json'
+import registryExchangesABI from '../abis/registry_exchanges.json'
+import erc20ABI from '../abis/ERC20.json'
+
+export const supportedEndpoints = ['crypto']
+
+export const endpointResultPaths = {
+  crypto: 'rate',
+}
+
+export interface ResponseSchema {
+  pool: string
+  input: string
+  inputToken: string
+  inputDecimals: number
+  output: string
+  outputToken: string
+  outputDecimals: number
+  rate: number
+}
+
+export const inputParameters: InputParameters = {
+  from: ['base', 'from', 'coin'],
+  fromAddress: false,
+  fromDecimals: false,
+  to: ['quote', 'to', 'market'],
+  toAddress: false,
+  toDecimals: false,
+  amount: false,
+  resultPath: false,
+}
+
+export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
+  const validator = new Validator(request, inputParameters)
+  if (validator.error) throw validator.error
+
+  const jobRunID = validator.validated.id
+  const { address: from, decimals: fromDecimals } = await getTokenDetails(validator, 'from', config)
+  const { address: to, decimals: toDecimals } = await getTokenDetails(validator, 'to', config)
+  const inputAmount = validator.validated.data.amount || 1
+  const amount = BigNumber.from(inputAmount).mul(BigNumber.from(10).pow(fromDecimals))
+  const resultPath = validator.validated.data.resultPath
+
+  const out = await getBestRate(from, to, amount, config)
+  const [pool, output] = out
+
+  const outputAmount = ethers.utils.formatUnits(output, toDecimals)
+  const rate = Number(outputAmount) / Number(inputAmount)
+
+  const data: ResponseSchema = {
+    pool,
+    input: amount.toString(),
+    inputToken: from,
+    inputDecimals: fromDecimals,
+    output: output.toString(),
+    outputToken: to,
+    outputDecimals: toDecimals,
+    rate,
+  }
+
+  const response = {
+    status: 200,
+    statusText: 'OK',
+    headers: {},
+    config: {},
+    data: data,
+  }
+  const result = Requester.validateResultNumber(response.data, [resultPath])
+
+  return Requester.success(jobRunID, Requester.withResult(response, result), config.verbose)
+}
+
+const getTokenDetails = async (
```

Comment:
How do you feel about adding some documentation for this function? Especially on all the OR conditionals that can occur and their ordering

Reply:
I agree, documentation here would be very valuable.